File: ImperatorToCK3/Source/Mappers/CultureMapper/CultureMapper.cpp
#include "CultureMapper.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"

mappers::CultureMapper::CultureMapper(std::istream& theStream)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

mappers::CultureMapper::CultureMapper()
{
	LOG(LogLevel::Info) << "-> Parsing culture mappings.";
	registerKeys();
	parseFile("configurables/culture_map.txt");
	clearRegisteredKeywords();
	LOG(LogLevel::Info) << "<> Loaded " << cultureMapRules.size() << " cultural links.";
}

void mappers::CultureMapper::registerKeys()
{
	registerKeyword("link", [this](std::istream& theStream) {
		const CultureMappingRule rule(theStream);
		cultureMapRules.push_back(rule);
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

std::optional<std::string> mappers::CultureMapper::match(const std::string& impCulture,
	 const std::string& ck3religion,
	 const unsigned long long ck3ProvinceID,
	 const unsigned long long impProvinceID,
	 const std::string& ck3ownerTitle) const
{
	for (const auto& cultureMappingRule: cultureMapRules)
	{
		const auto& possibleMatch = cultureMappingRule.match(impCulture, ck3religion, ck3ProvinceID, impProvinceID, ck3ownerTitle);
		if (possibleMatch)
			return *possibleMatch;
	}
	return std::nullopt;
}

std::optional<std::string> mappers::CultureMapper::nonReligiousMatch(const std::string& impCulture,
	const std::string& ck3religion,
	const unsigned long long ck3ProvinceID,
	const unsigned long long impProvinceID,
	const std::string& ck3ownerTitle) const
{
	for (const auto& cultureMappingRule : cultureMapRules)
	{
		const auto& possibleMatch = cultureMappingRule.nonReligiousMatch(impCulture, ck3religion, ck3ProvinceID, impProvinceID, ck3ownerTitle);
		if (possibleMatch)
			return *possibleMatch;
	}
	return std::nullopt;
}

void mappers::CultureMapper::loadRegionMappers(std::shared_ptr<ImperatorRegionMapper> impRegionMapper, std::shared_ptr<CK3RegionMapper> _ck3RegionMapper)
{
	const auto imperatorRegionMapper = std::move(impRegionMapper);
	const auto ck3RegionMapper = std::move(_ck3RegionMapper);
	if (!imperatorRegionMapper)
		throw std::runtime_error("Culture Mapper: Imperator Region Mapper is unloaded!");
	if (!ck3RegionMapper)
		throw std::runtime_error("Culture Mapper: CK3 Region Mapper is unloaded!");
	for (auto& mapping : cultureMapRules)
	{
		mapping.insertImperatorRegionMapper(imperatorRegionMapper);
		mapping.insertCK3RegionMapper(ck3RegionMapper);
	}
}

File: ImperatorToCK3/Mappers/Culture/CultureMapper.cs
using commonItems;
using ImperatorToCK3.CK3.Cultures;
using ImperatorToCK3.Mappers.Region;
using System.Collections.Generic;
using System.Linq;

namespace ImperatorToCK3.Mappers.Culture;

public sealed class CultureMapper {
	public CultureMapper(ImperatorRegionMapper irRegionMapper, CK3RegionMapper ck3RegionMapper, CultureCollection cultures) {
		this.irRegionMapper = irRegionMapper;
		this.ck3RegionMapper = ck3RegionMapper;

		Logger.Info("Parsing culture mappings...");
		var parser = new Parser();
		RegisterKeys(parser);
		parser.ParseFile("configurables/culture_map.txt");
		Logger.Info($"Loaded {cultureMappingRules.Count} cultural links.");
		
		RemoveInvalidRules(cultures);

		Logger.IncrementProgress();
	}
	public CultureMapper(BufferedReader reader, ImperatorRegionMapper irRegionMapper, CK3RegionMapper ck3RegionMapper, CultureCollection cultures) {
		this.irRegionMapper = irRegionMapper;
		this.ck3RegionMapper = ck3RegionMapper;

		var parser = new Parser();
		RegisterKeys(parser);
		parser.ParseStream(reader);
		
		RemoveInvalidRules(cultures);
	}

	private void RegisterKeys(Parser parser) {
		parser.RegisterKeyword("link", reader => cultureMappingRules.Add(CultureMappingRule.Parse(reader)));
		parser.IgnoreAndLogUnregisteredItems();
	}
	private void RemoveInvalidRules(CultureCollection cultures) {
		var validCultureIds = cultures.Select(c => c.Id);
		var removedCount = cultureMappingRules.RemoveAll(rule => !validCultureIds.Contains(rule.CK3CultureId));
		
		if (removedCount > 0) {
			Logger.Debug($"{removedCount} culture mapping rules removed due to specified CK3 cultures not existing.");
		}
	}
	
	public string? Match(
		string irCulture,
		ulong? ck3ProvinceId,
		ulong? irProvinceId,
		string? historicalTag
	) {
		foreach (var cultureMappingRule in cultureMappingRules) {
			var possibleMatch = cultureMappingRule.Match(irCulture, ck3ProvinceId, irProvinceId, historicalTag, irRegionMapper, ck3RegionMapper);
			if (possibleMatch is not null) {
				return possibleMatch;
			}
		}
		return null;
	}

	private readonly List<CultureMappingRule> cultureMappingRules = new();
	private readonly ImperatorRegionMapper irRegionMapper;
	private readonly CK3RegionMapper ck3RegionMapper;
}